### PR TITLE
[core] Fix the bug in task cancel hang.

### LINKED ISSUE #24359: [core] Ray task cancel race conditions
### What happened + What you expected to happen

Task cancelation is not handled correctly and will cause race conditions.

### Versions / Dependencies

master/linux

### Reproduction script

If we put 2 seconds sleep, test_cancel.py will hang:

```
diff --git a/python/ray/worker.py b/python/ray/worker.py
index 84ecaa277..33e836c57 100644
--- a/python/ray/worker.py
+++ b/python/ray/worker.py
@@ -1594,7 +1594,9 @@ def connect(

     # Notify raylet that the core worker is ready.
     worker.core_worker.notify_raylet()
-
+    if mode == WORKER_MODE:
+        from time import sleep
+        sleep(2)
     if driver_object_store_memory is not None:
         logger.warning(
             "`driver_object_store_memory` is deprecated"
```

```
pytest ray/tests/test_cancel.py::test_stress -vsx
```

The race condition is:
1. task created
2. task pushed to core worker
3. ray worker hasn't started to run the task
4. ray worker get the cancel
5. driver will hang then.

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
This PR fixes a bug: when the task is pushed to a core worker but hasn't been scheduled to run cancel is not called which will lead to the get request hanging forever.

The fix is to call the `Cancel`.

## Related issue number
Closes #24359 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -3035,23 +3035,24 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
...
     // If the task is not currently running, check if it is in the worker's queue of
     // normal tasks, and remove it if found.
     success = direct_task_receiver_->CancelQueuedNormalTask(task_id);
   }
   if (request.recursive()) {
     auto recursive_cancel = CancelChildren(task_id, request.force_kill());
-    if (recursive_cancel.ok()) {
-      RAY_LOG(INFO) << "Recursive cancel failed for a task " << task_id;
+    if (!recursive_cancel.ok()) {
+      RAY_LOG(ERROR) << "Recursive cancel failed for a task " << task_id;
```
Comment (mwtian) [inline src/ray/core_worker/core_worker.cc:3047]: Print the status as well?

File: src/ray/core_worker/transport/direct_task_transport.cc [hunk 2]
```diff
@@ -666,21 +666,32 @@ Status CoreWorkerDirectTaskSubmitter::CancelTask(TaskSpecification task_spec,
...
+                    boost::bind(&CoreWorkerDirectTaskSubmitter::CancelTask,
+                                this,
+                                task_spec,
+                                force_kill,
+                                recursive));
+              }
+            }
+          } else {
+            if (!reply.attempt_succeeded()) {
+              RAY_LOG(ERROR) << "Try to cancel task " << task_spec.TaskId()
```
Comment (mwtian) [inline src/ray/core_worker/transport/direct_task_transport.cc:692]: Nit: s/Try/Attempted

File: src/ray/core_worker/transport/direct_task_transport.cc [hunk 3]
```diff
@@ -666,21 +666,32 @@ Status CoreWorkerDirectTaskSubmitter::CancelTask(TaskSpecification task_spec,
...
+                       << " with status " << status.ToString();
         cancelled_tasks_.erase(task_spec.TaskId());
-
-        if (status.ok() && !reply.attempt_succeeded()) {
-          if (cancel_retry_timer_.has_value()) {
-            if (cancel_retry_timer_->expiry().time_since_epoch() <=
-                std::chrono::high_resolution_clock::now().time_since_epoch()) {
-              cancel_retry_timer_->expires_after(boost::asio::chrono::milliseconds(
-                  RayConfig::instance().cancellation_retry_ms()));
+        if (status.ok()) {
```
Comment (mwtian) [inline src/ray/core_worker/transport/direct_task_transport.cc:672]: Is it clearer to keep the `status.ok() && !reply.attempt_succeeded()` condition? We can remove one level of condition nesting in each branch.

File: src/ray/core_worker/transport/normal_scheduling_queue.cc [hunk 4]
```diff
@@ -68,6 +68,7 @@ bool NormalSchedulingQueue::CancelTaskIfFound(TaskID task_id) {
        it != pending_normal_tasks_.rend();
        ++it) {
     if (it->TaskID() == task_id) {
+      it->Cancel();
```
Comment (mwtian) [inline src/ray/core_worker/transport/normal_scheduling_queue.cc:71]: IIUC this is the fix?

**Review by mwtian — COMMENTED** [review]
Good catch!

File: src/ray/core_worker/transport/normal_scheduling_queue.cc (hunk 4 above)
Comment (iycheng) [reply src/ray/core_worker/transport/normal_scheduling_queue.cc:71]: Yes

**Review by mwtian — APPROVED** [review]

File: src/ray/core_worker/transport/direct_task_transport.cc [hunk 5]
```diff
@@ -666,25 +666,41 @@ Status CoreWorkerDirectTaskSubmitter::CancelTask(TaskSpecification task_spec,
...
+                       << " with status " << status.ToString();
         cancelled_tasks_.erase(task_spec.TaskId());
 
-        if (status.ok() && !reply.attempt_succeeded()) {
-          if (cancel_retry_timer_.has_value()) {
-            if (cancel_retry_timer_->expiry().time_since_epoch() <=
-                std::chrono::high_resolution_clock::now().time_since_epoch()) {
-              cancel_retry_timer_->expires_after(boost::asio::chrono::milliseconds(
-                  RayConfig::instance().cancellation_retry_ms()));
+        // Retry is not attempted if !status.ok() because force-kill may kill the worker
```
Comment (scv119) [inline src/ray/core_worker/transport/direct_task_transport.cc:673]: just to clarify. this doesn't change the logic but just move the code around right?

**Review by scv119 — APPROVED** [review]

File: src/ray/core_worker/transport/direct_task_transport.cc (hunk 5 above)
Comment (iycheng) [reply src/ray/core_worker/transport/direct_task_transport.cc:673]: No logic change. Just more logs for error case.

**Pull request merged**